core/src/main/java/dk/alexandra/fresco/framework/network/async/AsyncNetwork.java

```diff
@@ -0,0 +1,336 @@
+  }
+
+  /**
+   * Creates a network with the given configuration and a timeout of <code>timeout</code> in
+   * milliseconds. Calling the constructor will automatically trigger an attempt to connect to the
+   * other parties.
+   *
+   * @param conf The network configuration
+   * @param timeout the time to wait until timeout
+   */
+  public AsyncNetwork(NetworkConfiguration conf, Duration timeout) {
+    this.conf = conf;
+    this.queues = new HashMap<>(conf.noOfParties());
+    this.channelMap = new ConcurrentHashMap<>(conf.noOfParties() - 1);
+    for (int i = 1; i < conf.noOfParties() + 1; i++) {
+      queues.put(i, new LinkedBlockingQueue<>());
+    }
+    if (conf.noOfParties() > 1) {
+      connectNetwork(conf, timeout);
+      startReceivers();
+    }
+    logger.info("P{} successfully connected network", conf.getMyId());
+  }
+
+  /**
+   * Fully connects the network.
+   * <p>
+   * Connects two channels to each external party (i.e., parties other than this party). One channel
+   * to send and one channel to receive messages.
+   * </p>
+   *
+   * @param conf the configuration defining the network to connect
+   * @param timeout duration to wait until timeout
+   */
+  private void connectNetwork(NetworkConfiguration conf, Duration timeout) {
+    int externalParties = getNoOfParties() - 1;
+    this.receiverService = Executors.newFixedThreadPool(externalParties);
+    this.senderServices = new HashMap<>(externalParties);
+    for (int i = 1; i <= conf.noOfParties(); i++) {
+      if (this.conf.getMyId() != i) {
+        this.senderServices.put(i, Executors.newSingleThreadExecutor());
+      }
+    }
+    ExecutorService es = Executors.newFixedThreadPool(2);
+    CompletionService<Map<Integer, SocketChannel>> cs = new ExecutorCompletionService<>(es);
+    cs.submit(() -> {
+      return connectClient();
+    });
+    cs.submit(() -> {
+      bindServer();
+      return connectServer();
+    });
+    try {
+      for (int i = 0; i < 2; i++) {
+        Future<Map<Integer, SocketChannel>> f = cs.poll(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        if (f == null) {
+          throw new TimeoutException("Timed out");
+        } else {
+          this.channelMap.putAll(f.get());
+        }
+      }
+    } catch (InterruptedException e) {
+      close();
+      throw new RuntimeException("Interrupted while connecting network", e);
+    } catch (ExecutionException e) {
+      close();
+      throw new RuntimeException("Failed to connect network", e.getCause());
+    } catch (TimeoutException e) {
+      close();
+      throw new RuntimeException("Timed out connecting network", e);
+    } finally {
+      es.shutdownNow();
+    }
+  }
+
+  private void startReceivers() {
+    for (Entry<Integer, SocketChannel> entry : this.channelMap.entrySet()) {
+      final int id = entry.getKey();
+      SocketChannel channel = entry.getValue();
+      this.receiverService.submit(() -> {
+        try {
+          while (true) {
```

Comment:
while(true) and executorService is a poor match, either use a thread or resubmit. 
This would also change the structure of the exception propagation

Reply:
How would it work with resubmits? I cache the future and check if the latest future was cancelled in the receive method?

Reply:
Well in the receive you only schedule the next receive if the current was successful and subsequently overwrite the "next" future.

Reply:
Will that work? Do I have a guarantee that the last receive is done when I submit the next?

Reply:
In this case (recieving) there is only one entity submitting, so yes